packages/fiber/src/core/index.tsx

```diff
@@ -0,0 +1,405 @@
+      const outputEncoding = linear ? THREE.LinearEncoding : THREE.sRGBEncoding
+      const toneMapping = flat ? THREE.NoToneMapping : THREE.ACESFilmicToneMapping
+      if (gl.outputEncoding !== outputEncoding) gl.outputEncoding = outputEncoding
+      if (gl.toneMapping !== toneMapping) gl.toneMapping = toneMapping
+
+      // Set gl props
+      if (glConfig && !is.fun(glConfig) && !isRenderer(glConfig) && !is.equ(glConfig, gl, shallowLoose))
+        applyProps(gl as any, glConfig as any)
+      // Store events internally
+      if (events && !state.events.handlers) state.set({ events: events(store) })
+      // Check pixelratio
+      if (dpr && state.viewport.dpr !== calculateDpr(dpr)) state.setDpr(dpr)
+      // Check size, allow it to take on container bounds initially
+      size = size || { width: canvas.parentElement?.clientWidth ?? 0, height: canvas.parentElement?.clientHeight ?? 0 }
+      if (!is.equ(size, state.size, shallowLoose)) state.setSize(size.width, size.height)
+      // Check frameloop
+      if (state.frameloop !== frameloop) state.setFrameloop(frameloop)
+      // Check pointer missed
+      if (!state.onPointerMissed) state.set({ onPointerMissed })
+      // Check performance
+      if (performance && !is.equ(performance, state.performance, shallowLoose))
+        state.set((state) => ({ performance: { ...state.performance, ...performance } }))
+
+      // Set locals
+      onCreated = onCreatedCallback
+      configured = true
+
+      return this
+    },
+    render(element: React.ReactNode) {
+      // The root has to be configured before it can be rendered
+      if (!configured) this.configure()
+
+      reconciler.updateContainer(
+        <Provider store={store} element={element} onCreated={onCreated} target={canvas} />,
+        fiber,
+        null,
+        () => undefined,
+      )
+      return store
+    },
+    unmount() {
+      unmountComponentAtNode(canvas)
+    },
+  }
+}
+
+function render<TCanvas extends Element>(
+  element: React.ReactNode,
+  canvas: TCanvas,
+  config: RenderProps<TCanvas> = {},
+): UseBoundStore<RootState> {
+  console.warn('R3F.render is no longer supported in React 18. Use createRoot instead!')
+  const root = createRoot(canvas)
+  root.configure(config)
+  return root.render(element)
+}
+
+function Provider<TElement extends Element>({
+  store,
+  element,
+  onCreated,
+  target,
+}: {
+  onCreated?: (state: RootState) => void
+  store: UseBoundStore<RootState>
+  element: React.ReactNode
+  target: TElement
+}) {
+  React.useEffect(() => {
+    const state = store.getState()
+    // Flag the canvas active, rendering will now begin
+    state.set((state) => ({ internal: { ...state.internal, active: true } }))
+    // Connect events to the targets parent, this is done to ensure events are registered on
+    // a shared target, and not on the canvas itself
+    state.events.connect?.(target.parentNode)
```

Comment:
This is tripping up the CI. I'll mock the original behavior in RTTR's canvas polyfill.

Reply:
it's not good either way because parentNode is dom specific. the problem is this ...

i need events on the element that wraps the canvas node. easy for our `<Canvas>` components because it has a container div, but the render api doesnt have any: `createRoot(canvas).render(<App />)`. split view portals cannot not work otherwise. and ideas?

Reply:
i hope its a passable solution now. canvas will have to register its parent, which defines the source not only for events but also for appended elements (html, views, scrollcontrols, etc). still need to write it into the state model so that drei can use it.